feat: ✨ lock member default availability

shaun-ak opened the pull request:
## What does this PR do?


- Fixes #22089 (GitHub issue number)
- Fixes CAL-5995 (Linear issue number - should be visible at the bottom of the GitHub issue description)

#### Video Demo (if applicable):

https://github.com/user-attachments/assets/eacb5255-0ec7-4f30-9da8-0ad4fb18b37d

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

    

<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Added an organization-only team setting to lock default availability, preventing members from editing, deleting, or changing the timezone of their default schedule unless they are an admin or owner. Members can still create additional schedules.

- **New Features**
  - UI toggle in Team Settings (admins only; shown for teams within organizations).
  - When locked, default schedule actions are disabled with clear error/toast and a warning banner.
  - Timezone updates skip the default schedule when locked (profile updates and cron timezone changes).

<sup>Written for commit 923b7b2894fb8d12df5298ab2ee61b08f23c6be2. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

shaun-ak committed:
feat: lock member default availability

Devanshusharma2005 reviewed (changes requested):
Hey can you please address the issues suggested by the cubic

shaun-ak commented on packages/prisma/schema.prisma:
```diff
@@ -464,70 +464,71 @@ model Profile {
+  members                 Membership[]
+  eventTypes              EventType[]
+  workflows               Workflow[]
+  createdAt               DateTime                @default(now())
```
now() is being used extensively in schema.prisma.

shaun-ak committed:
address cubic pr comments

shaun-ak commented on packages/lib/schedules/updateSchedule.ts:
```diff
@@ -59,6 +60,41 @@ export const updateSchedule = async ({ input, user, prisma }: IUpdateScheduleOpt
+      select: {
+        team: {
+          select: {
+            id: true,
```
removed them

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

Devanshusharma2005 reviewed (commented):
okay . Lets wait for @anikdhabal review too.

kart1ka reviewed (commented):
Left a comment. Need clarification from the team. The rest of the code looks good to me.

shaun-ak commented on packages/trpc/server/routers/viewer/availability/schedule/create.handler.ts:
```diff
@@ -61,6 +62,34 @@ export const createHandler = async ({ input, ctx }: CreateOptions) => {
   });
 
   if (!user.defaultScheduleId) {
+    // Check if user is a member of any team with locked default availability
+    const userTeams = await prisma.membership.findMany({
+      where: {
+        userId: user.id,
+        accepted: true,
+      },
+      select: {
+        team: {
+          select: {
+            lockDefaultAvailability: true,
+          },
+        },
+        role: true,
+      },
+    });
+
+    // Check if user is a member (not admin/owner) of any team with locked default availability
+    const hasLockedTeamMembership = userTeams.some(
+      (membership) => membership.team.lockDefaultAvailability && membership.role === MembershipRole.MEMBER
+    );
+
+    if (hasLockedTeamMembership) {
+      throw new TRPCError({
+        code: "FORBIDDEN",
+        message: "Cannot edit default availability when team has locked default availability setting enabled",
+      });
+    }
+
```
I have 1 more doubt, so you can see I used `some` here. 

So, if a user is a member in one team (with locked default availability) but owner/admin in another then should we allow him/her to do CRUD on default schedules or not, because some will take account of the team where the user is a member and has locked default availability.

kart1ka replied:
will check with the team on this.

shaun-ak replied:
> @shaun-ak A team or org admin/owner should always be able to update their own and their team members’ default availabilities. Even if they’re just a member in another team, they should still be able to change their own default availability.

I have also pushed logic for this 7125a3c96418378188e7615dd5e0d850ad257a3d

kart1ka reviewed (changes requested):
In a team inside an org, when a team admin/owner tries to update their own default availability, no error is shown. Default Availability does not update but on the frontend no error is shown.

https://cap.link/sca2b78tdqjecpr

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: add error message in availability edit sheet

kart1ka commented:
> Hey @kart1ka , I used the seeded users for acme, but not able to edit the availabilty, am I missing something--
> 
> [cap.so/s/faf4dkfhhsab8sc](https://cap.so/s/faf4dkfhhsab8sc)

Team and Org Admin and Owners can edit other user's availabilities. I do not think member 1 is an admin/owner of any team.

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented:
> In a team inside an org, when a team admin/owner tries to update their own default availability, no error is shown. Default Availability does not update but on the frontend no error is shown.
> 
> https://cap.link/sca2b78tdqjecpr

Hey @kart1ka , if a team admin/owner tries to update their own default availability, then they should be able to do it.
Default availability lock is only for members of a team.

https://cap.link/qtwxky6ykk2e238 (I made member 1 as admin and was able to update availability)

shaun-ak committed:
feat: fix owner or admin based access for availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka commented on packages/lib/schedules/updateSchedule.ts:
```diff
@@ -59,6 +60,46 @@ export const updateSchedule = async ({ input, user, prisma }: IUpdateScheduleOpt
   }
 
+  // Check if this is the user's default schedule and if they are a member of any team with locked default availability
+  const isDefaultSchedule = user.defaultScheduleId === input.scheduleId;
+  const isSettingAsDefault = input.isDefault;
+
+  if (isDefaultSchedule || isSettingAsDefault) {
+    // Check if user is a member of any team with locked default availability
+    const userTeams = await prisma.membership.findMany({
+      where: {
+        userId: user.id,
+        accepted: true,
+      },
+      select: {
+        team: {
+          select: {
+            lockDefaultAvailability: true,
+          },
+        },
+        role: true,
+      },
+    });
+
+    // Check if user is a member (not admin/owner) of any team with locked default availability
+    const hasLockedTeamMembership = userTeams.some(
+      (membership) => membership.team.lockDefaultAvailability && membership.role === MembershipRole.MEMBER
+    );
+
+    // Check if user is an admin/owner of any team (which gives them permission to override restrictions)
+    const hasAdminOrOwnerRole = userTeams.some(
+      (membership) => membership.role === MembershipRole.ADMIN || membership.role === MembershipRole.OWNER
+    );
+
+    // Only block if user has locked team membership AND is not an admin/owner of any team
+    if (hasLockedTeamMembership && !hasAdminOrOwnerRole) {
+      throw new TRPCError({
+        code: "FORBIDDEN",
+        message: "Cannot edit default availability when team has locked default availability setting enabled",
+      });
+    }
```
Since this logic is being repeated across multiple places, it would be better to abstract it into a reusable function and call that function instead.

kart1ka commented on packages/trpc/server/routers/viewer/availability/schedule/create.handler.ts:
```diff
@@ -61,6 +62,40 @@ export const createHandler = async ({ input, ctx }: CreateOptions) => {
   });
 
   if (!user.defaultScheduleId) {
+    // Check if user is a member of any team with locked default availability
+    const userTeams = await prisma.membership.findMany({
+      where: {
+        userId: user.id,
+        accepted: true,
+      },
+      select: {
+        team: {
+          select: {
+            lockDefaultAvailability: true,
+          },
+        },
+        role: true,
+      },
+    });
+
+    // Check if user is a member (not admin/owner) of any team with locked default availability
+    const hasLockedTeamMembership = userTeams.some(
+      (membership) => membership.team.lockDefaultAvailability && membership.role === MembershipRole.MEMBER
+    );
+
+    // Check if user is an admin/owner of any team (which gives them permission to override restrictions)
+    const hasAdminOrOwnerRole = userTeams.some(
+      (membership) => membership.role === MembershipRole.ADMIN || membership.role === MembershipRole.OWNER
+    );
+
+    // Only block if user has locked team membership AND is not an admin/owner of any team
+    if (hasLockedTeamMembership && !hasAdminOrOwnerRole) {
+      throw new TRPCError({
+        code: "FORBIDDEN",
+        message: "Cannot edit default availability when team has locked default availability setting enabled",
+      });
+    }
+
```
I do not think we should prevent user from creating a default availability if it already does not exist. Let me check with the team on that.

shaun-ak replied:
I think there will always be a default availability right?

The current code ensures that at least one availabilty schedule exists and that would be set as the default one.

kart1ka reviewed (changes requested):
Left a few suggestions.
Also, When the member adds date override to the default availability, it adds on the frontend and a success toast is shown in the ui. This should not be the case. 
See - https://cap.so/s/c18kc2mhzk35y93

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: add restriction on date override

shaun-ak commented:
Hey @kart1ka , I have made changes to disable "add override" button and also added alert banner and diasbled the schedule listing if default availability is locked.

<img width="1034" height="806" alt="Screenshot 2025-07-17 at 8 55 45 AM" src="https://github.com/user-attachments/assets/aad43f08-5812-4f90-8e7d-d4a0139e4015" />

Devanshusharma2005 reviewed (changes requested):
hey @shaun-ak the typecheck issues are failing. Please try to address them. Marking it draft until then. Feel free to rfr.

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
fix: failing type check

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'feat-lock-member-default-availability' of https://github.com/shaun-ak/cal.com into feat-lock-member-default-availability

shaun-ak commented:
Fixed the type check issues @Devanshusharma2005

shaun-ak committed:
feat: cleanup code

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka reviewed (changes requested):
Pls also disable timezone field and delete button in default availability.

Getting the following error when editing the default availability of team member. Can you please fix that? Not getting this error on main.

<img width="1032" height="1660" alt="Image" src="https://github.com/user-attachments/assets/ec39639e-32c7-46b3-bb46-20d9e1d0bd36" />

shaun-ak committed:
feat: disable timezone select and delete button

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka commented on packages/lib/lockedDefaultAvailability.ts:
```diff
@@ -0,0 +1,59 @@
+ */
+export async function hasLockedDefaultAvailabilityRestriction(
+  userId: number,
+  prisma: PrismaClient
```
i don't think we need to pass prisma as an argument. we can just import it.

kart1ka commented on packages/lib/lockedDefaultAvailability.ts:
```diff
@@ -0,0 +1,59 @@
+ * This is a convenience function that combines the check with error throwing.
+ */
+export async function checkLockedDefaultAvailabilityRestriction(
+  userId: number,
+  prisma: PrismaClient,
+  errorMessage = "Cannot edit default availability when team has locked default availability setting enabled"
```
Same here. we don't need to pass prisma and errorMessage as arguments.

kart1ka reviewed (changes requested):
Left a few comments.

shaun-ak committed:
feat: remove error message and prisma as arguments

shaun-ak committed:
feat: fix type check error

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

CarinaWolli reviewed (changes requested):
Nice work so far 🙏 I found a few issues that need to be addressed first: 

- [ ] When profile timezone is updated, we update the default schedule timezone automatically. We shouldn't do this when locking availability is enabled 
- [ ] Same also for travel schedules
- [ ] Let's also disable or remove everything the member can't use. There might be more, but that's what I found:
 
Disable save button
<img width="372" height="87" alt="Screenshot 2025-07-23 at 10 34 48 AM" src="https://github.com/user-attachments/assets/5d17ce02-2586-436a-bbf4-d85bc85642bd" />

Remove 'Delete' in the dropdown
<img width="372" height="184" alt="Screenshot 2025-07-23 at 10 25 34 AM" src="https://github.com/user-attachments/assets/bf837d2a-c998-47b0-83e0-1788b64a46b3" />

Remove 'Set as default' from the dropdown
<img width="314" height="186" alt="Screenshot 2025-07-23 at 10 26 31 AM" src="https://github.com/user-attachments/assets/b497b7a0-bd59-44c8-b26c-d51a89addae4" />

shaun-ak committed:
Merge remote-tracking branch 'origin' into feat-lock-member-default-availability

shaun-ak committed:
feat: fix merge conflict issue

shaun-ak committed:
feat: lock default timezone updation if default availability is locked

shaun-ak committed:
feat: lock default timezone for travel schedule update

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: remove options from dropdown

shaun-ak committed:
feat: fix type check issue

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka committed:
fix: locked availability font size

kart1ka commented:
> Thanks for the detailed feedback, @CarinaWolli. I've addressed all the points and also disabled the "Set to default" switch for non-default schedules.
> 
> Let me know if there's anything else you'd like to tweak!
> 
> Here's a video recording of the changes -- [cap.link/kmthe6t070472r7](https://cap.link/kmthe6t070472r7)

When profile timezone is updated, the default timezone is also updated when locked default availability is on.
See the video: https://cap.link/31knrngrzp224vh

Pls make sure the changes work for both profile timezone update and the travel schedule update.

kart1ka reviewed (changes requested):
left a comment

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented:
https://github.com/user-attachments/assets/dd744996-5538-4dc5-8641-229ee0a7b0e7

Hey @kart1ka , for me the scenario is working fine, was there any specific testing path that you followed?

Attached a video of my local testing.

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka reviewed (changes requested):
Everything seems to be working fine. Can you please add some tests to verify the logic works?

shaun-ak committed:
feat: add test cases for default-availability file

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented:
Hey @kart1ka , I have added test cases for lockedDefaultAvailability file. commit-- 318cccfc7de16973d946f300c80f741d8165ce1b

CarinaWolli commented on apps/web/app/api/cron/changeTimeZone/route.ts:
```diff
@@ -40,28 +41,33 @@ async function postHandler(request: NextRequest) {
-    if (!user.defaultScheduleId) {
-      // set default schedule if not already set
-      await prisma.user.update({
+    if (!hasLockedAvailability) {
```
I think we also need to handle the same case when updating the profile timezone via API v2

shaun-ak replied:
Sure @CarinaWolli , looking into it.

shaun-ak replied:
Hey @CarinaWolli , in v2 if we update the profile timezone using the `/me` patch then the default availability timezone is not updated, only the profile timezone gets updated.

CarinaWolli replied:
Thanks for checking, then it should be fine 👍

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

CarinaWolli reviewed (changes requested):
Hey @shaun-ak, everything works great. One more thing: we just decided to make this feature available only for organizations. It’s still a team-wide setting, but we want to offer it only for teams within an organization. Could you add that as well? 🙏

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented:
Sure @CarinaWolli , will make it work for only organizations.

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: lock default availability only for organization teams

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
fix: merge issues

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented on packages/platform/atoms/availability/AvailabilitySettings.tsx:
```diff
@@ -269,7 +269,12 @@ const DateOverride = ({
           Trigger={
-            <Button color="secondary" StartIcon="plus" data-testid="add-override" disabled={disabled}>
+            <Button
+              className={classNames?.button}
```
Git is showing this as changes made by me but these actually came from resolving merg conflicts.

I just added `disabled` to disable the date over ride button.

[PR that was merged added this piece of code](https://github.com/calcom/cal.com/pull/22976/files#diff-9141f5df2ef463aeee7fe93f889d1f617927d9901bbfa61f882576dcce7cd995R254)

shaun-ak commented on packages/platform/atoms/availability/AvailabilitySettings.tsx:
```diff
@@ -712,19 +717,22 @@ export const AvailabilitySettings = forwardRef<AvailabilitySettingsFormRef, Avai
+                    disabled={schedule.isDefault && schedule.lockedDefaultAvailability}
+                    classNames={customClassNames?.dateOverrideClassNames}
+                  />
+                </BookerStoreProvider>
```
Same for this too, [PR that was merged for this piece of code](https://github.com/calcom/cal.com/pull/22925/files#diff-9141f5df2ef463aeee7fe93f889d1f617927d9901bbfa61f882576dcce7cd995R676).

I just added `disabled` that is passed to `dateOverride` function.

shaun-ak commented on packages/prisma/schema.prisma:
```diff
@@ -551,9 +551,11 @@ model Team {
+  domainWideDelegations  DomainWideDelegation[]
+  roles                  Role[] // Added for Role relation
+  calAiPhoneNumbers      CalAiPhoneNumber[]
+  agents                 Agent[]
```
Same here too, [Piece of code that was merged](https://github.com/calcom/cal.com/pull/22919/files#diff-56fa4d384ba6fb94b130e9eb42c5cad00734b025ebd6c197f76873469725ae87R444)

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

CarinaWolli reviewed (dismissed):
Nice work, tested it and it's working as expected 🙌🏻

This still needs review from @calcom/dba 🙏

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

volnei reviewed (changes requested):
Nit comments but would be good to check

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
fix: address pr review comments

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

kart1ka committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: changes to fix timezone setting in schedules

shaun-ak commented:
I have made the fix for showing timeZone of schedule and then using user.timeZone as a fallback.

https://github.com/user-attachments/assets/ba847456-be27-4e03-b60a-6d61aa6ae66d

CC: @keithwillcode @volnei @CarinaWolli

shaun-ak committed:
feat: merge main into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: fix type check issue

shaun-ak commented:
Hey @keithwillcode @volnei , please review this PR.

I have tested it in my local and it is working as expected.

shaun-ak commented:
Hey @keithwillcode @volnei , please review this PR.

I have tested it in my local and it is working as expected.

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

dhairyashiil reviewed (changes requested):
Great work on this PR! The PR is very close to being ready for merge. 

But PR violates our architecture pattern where all database queries must go through repository classes.

dhairyashiil commented on packages/lib/lockedDefaultAvailability.ts:
```diff
@@ -0,0 +1,52 @@
+import prisma from "@calcom/prisma";
```
1. Create a repository method
2. then use that here instead of Prisma directly

shaun-ak replied:
@dhairyashiil I have made the changes.

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak committed:
feat: use repository and remove any

shaun-ak committed:
feat: update profile update logic

shaun-ak committed:
feat: update availability schedules when timezone is changed

shaun-ak committed:
fix: type issue

shaun-ak committed:
fix: type issue

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

# Conflicts:
#	apps/web/public/static/locales/en/common.json
#	packages/features/schedules/repositories/ScheduleRepository.ts
#	packages/lib/schedules/updateSchedule.ts

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

shaun-ak commented:
https://github.com/user-attachments/assets/1df475c7-6c84-498e-8664-c17c4083f1eb

shaun-ak committed:
Merge branch 'main' into feat-lock-member-default-availability

pallava-joshi committed:
Merge branch 'main' into feat-lock-member-default-availability

pallava-joshi reviewed (commented):
can you please check on cubic suggestions too

devin-ai-integration[bot] committed:
Merge upstream/main into feat-lock-member-default-availability

Resolved conflicts in:
- apps/web/app/api/cron/changeTimeZone/route.ts
- apps/web/public/static/locales/en/common.json
- packages/features/schedules/components/ScheduleListItem.tsx
- packages/features/schedules/repositories/ScheduleRepository.ts
- packages/platform/atoms/availability/AvailabilitySettings.tsx
- packages/trpc/server/routers/viewer/availability/list.handler.ts
- packages/trpc/server/routers/viewer/availability/schedule/delete.handler.ts
- packages/trpc/server/routers/viewer/me/updateProfile.handler.ts

Conflict resolutions:
- Updated imports to use new ScheduleRepository from features package
- Updated HostRepository import to new location in features package
- Preserved locked default availability feature while adopting upstream changes
- Combined PR's locked availability checks with upstream's dialog-based delete confirmation
- Merged upstream's formHasChanges disabled state with PR's locked availability disabled state

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: address PR review feedback for lock member default availability

- Fix small-screen 'Set to default' switch disabled condition by removing
  redundant 'value' check that prevented toggling when already checked
- Fix effectiveParentId check in teams API to consider data.parentId when
  setting lockDefaultAvailability in the same request
- Fix timezone change guard in updateProfile to only trigger when timeZone
  is explicitly provided in input, preventing false positives

Co-authored-by: shaun-ak <shaun-ak@users.noreply.github.com>
Co-Authored-By: unknown <>

sahitya-chandra reviewed (commented):
Thank you for your contribution. This PR is being closed because it modifies code that no longer exists in Cal.diy
